### PR TITLE
Allow disabling PQ-based QP cache rewarm when reloading schemas

### PR DESCRIPTION
**Pull request by glasser**
In #3829 we taught router to use the PQ list to prewarm the query planner cache when reloading its schema.

In #5340 we gave it the experimental option to do the same thing at startup.

Now we are allowing you to turn off the on-reload prewarm.

Because the PQ list does not have metadata to tell the Router which PQs have been used recently, if you have a large PQ list then the on-reload prewarm can slow down schema updates noticeably, and it may not provide much value over prewarming based on the contents of the in-memory QP cache itself.

We keep the same defaults, but let you turn off the on-reload prewarm with:

```
persisted_queries:
  experimental_prewarm_query_plan_cache:
    on_reload: false
```

Previously, `persisted_queries.experimental_prewarm_query_plan_cache` was a boolean; a migration moves this boolean to
`persisted_queries.experimental_prewarm_query_plan_cache.on_startup`. (`on_startup` defaults to false and `on_reload` defaults to true.)

This migration uses a `type: change` action. No such actions existed in the codebase and the implementation appeared to be buggy: there is no top-level `==` operator (or top-level operators at all) in JSONPath. This PR fixes the implementation of `type: change` to use the `[?(@.x == y)]` filter syntax which does appear to work (as tested by the migration tests for the new migration).

This also renames a migration file added in #5957 that used a non-standard filename.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

I'm not sure where to put tests of the prewarming functionality; neither #3829 nor #5340 appear to have tests. (The migration is tested.)

**Exceptions**

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/query_planner/caching_query_planner.rs [hunk 1]
```diff
@@ -223,8 +224,9 @@ where
 
         cache_keys.shuffle(&mut thread_rng());
 
-        let should_warm_with_pqs =
-            (experimental_pql_prewarm && previous_cache.is_none()) || previous_cache.is_some();
+        let should_warm_with_pqs = (experimental_pql_prewarm.on_startup
+            && previous_cache.is_none())
+            || (experimental_pql_prewarm.on_reload && previous_cache.is_some());
```
Comment (glasser) [inline apollo-router/src/query_planner/caching_query_planner.rs:229]: This line is the heart of the PR.

File: apollo-router/src/services/layers/persisted_queries/manifest_poller.rs [hunk 2]
```diff
@@ -783,15 +782,14 @@ mod tests {
...
-                    Some(true),
-                    Some(false),
-                    Some(PersistedQueriesSafelist::default()),
-                    Some(false),
-                    Some(vec![
-                        "tests/fixtures/persisted-queries-manifest.json".to_string()
-                    ]),
-                ))
+                .persisted_query(
+                    PersistedQueries::builder()
```
Comment (glasser) [inline apollo-router/src/services/layers/persisted_queries/manifest_poller.rs:786]: Switching to using the builder seemed more legible if I had to update this anyway.

**Review by lleadbet — APPROVED** [review]
this looks great, thanks for tackling this so quickly!

File: apollo-router/src/configuration/upgrade.rs [hunk 3]
```diff
@@ -144,7 +144,7 @@ fn apply_migration(config: &Value, migration: &Migration) -> Result<Value, Confi
                 }
             }
             Action::Change { path, from, to } => {
-                if !jsonpath_lib::select(config, &format!("$.{path} == {from}"))
```
Comment (glasser) [inline apollo-router/src/configuration/upgrade.rs:147]: FWIW this syntax just seemed to be wrong (and the library ignored everything after the space without an error?)

https://freestrings.github.io/jsonpath/ can be used to test this jsonpath_lib implementation.

File: docs/source/configuration/in-memory-caching.mdx [hunk 4]
```diff
@@ -63,6 +61,8 @@ supergraph:
     warmed_up_queries: 100
 ```
 
+(In addition, the router can use the contents of the [persisted query list](./persisted-queries) to prewarm the cache. By default, it does this when loading a new schema but not on startup; you can [configure](./persisted-queries#persisted-queries#experimental_prewarm_query_plan_cache) it to change either of these defaults.)
```
Comment (chandrikas) [inline docs/source/configuration/in-memory-caching.mdx:64]: I think I might have gotten this wrong from the slack convo, but aren't we changing the default to be "warm PQs at startup" with an additional option to "warm PQs at schema reload"?

**Review by Geal — APPROVED** [review]

File: docs/source/configuration/in-memory-caching.mdx (hunk 4 above)
Comment (glasser) [reply docs/source/configuration/in-memory-caching.mdx:64]: (although the option's *name* is changing, with migration)

**Pull request merged**